[GStreamer][WebCodecs] HEVC encoding and decoding support

philn opened the pull request:
#### 904e4fcfa2e26ad5bfc0d4f9f7e7dfc4dde943cc
<pre>
[GStreamer][WebCodecs] HEVC encoding and decoding support
<a href="https://bugs.webkit.org/show_bug.cgi?id=260124">https://bugs.webkit.org/show_bug.cgi?id=260124</a>

Reviewed by Xabier Rodriguez-Calvar.

The corresponding layout tests are now passing, excepted the h265_annexb variants, that would
require further investigation. Support for H265 codec string to GStreamer H265 profile was also
added, with unit-tests.

* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/GStreamer.cmake:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::configureVideoDecoderForHarnessing):
(WebCore::wrapSpanData):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::initializeEncoders):
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::initialize):
* Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.cpp:
(WebCore::GStreamerCodecUtilities::parseHEVCProfile):
* Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.h:
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
(webkit_video_encoder_class_init):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/266904@main">https://commits.webkit.org/266904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3d31a22ff593efd90db0742f969ae445fb7b36f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15398 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15750 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16848 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14188 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15231 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17914 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15497 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16821 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15274 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15737 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12830 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17576 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13017 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13616 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20581 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12926 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14097 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13784 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17030 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14347 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14344 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12152 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15288 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13625 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3899 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3633 "layout-tests (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17962 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15521 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1821 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14185 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3709 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -7396,8 +7396,10 @@ WebCodecsHEVCEnabled:
+      "USE(GSTREAMER)": true
       default: false
     WebCore:
+      "USE(GSTREAMER)": true
```
Why this? Is this a debugging leftover?

philn replied:
This setting needs to be on for HEVC support to be enabled, it's not debugging leftover.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -1048,6 +1054,20 @@ bool gstObjectHasProperty(GstPad* pad, const char* name)
+
+    auto bufferSize = data.size();
+    auto bufferData = data.data();
+    auto buffer = adoptGRef(gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, bufferData, bufferSize, 0, bufferSize, new Vector<uint8_t>(WTFMove(data)), [](gpointer data) {
```
I think you can check above for the span to be empty and then create here the vector directly from the span.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.cpp:
```diff
@@ -66,6 +67,31 @@ std::pair<const char*, const char*> GStreamerCodecUtilities::parseH264ProfileAnd
+    for (unsigned i = 5, j = 0; i < 10; i++, j++)
+        profileTierLevel[i] = constraints[j];
+
+    return gst_codec_utils_h265_get_profile(profileTierLevel, sizeof(profileTierLevel));
```
Since you're returning the same thing just after the constain copy, I would put the constain copy inside an `if (profileTierLevel[0] >= 4)` and then return after it.

calvaris commented on Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
```diff
@@ -840,6 +841,122 @@ static void webkit_video_encoder_class_init(WebKitVideoEncoderClass* klass)
+                        supports12BitsBigEndian = true;
+                }
+
+                auto profile = makeString(profileString);
```
This can maybe a StringView

StaZhu commented on Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
```diff
@@ -66,6 +66,43 @@ TEST_F(GStreamerTest, gstStructureJSONSerializing)
+    ASSERT_STREQ(parseHEVCProfile("hev1.1.6.L93.B0"_s), "main");
+    ASSERT_STREQ(parseHEVCProfile("hev1.2.4.L93.B0"_s), "main-10");
+    ASSERT_STREQ(parseHEVCProfile("hev1.3.E.L93.B0"_s), "main-still-picture");
+    ASSERT_STREQ(parseHEVCProfile("hev1.4.16.L186.BF.C8"_s), "monochrome");
```
Per `ISO-14496-15Y2019 E3`:

```
the 32 bits of the general_profile_compatibility_flags, but in reverse bit order, i.e. with general_profile_compatibility_flag[ 31 ] as the most significant bit, followed by general_ profile_compatibility_flag[ 30 ], and down to general_profile_compatibility_flag[ 0 ] as the least significant bit, where general_profile_compatibility_flag[ i ] for i in the range of 0 to 31, inclusive, are specified in ISO/IEC 23008-2, encoded in hexadecimal (leading zeroes may be omitted)
```

Thus `general_profile_compatibility_flags` should be encoded in `hexadecimal` not `decimal`, so I guess you need to replace this to `hev1.4.10.L186.BF.C8`

StaZhu commented on Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
```diff
@@ -66,6 +66,43 @@ TEST_F(GStreamerTest, gstStructureJSONSerializing)
+    ASSERT_STREQ(parseHEVCProfile("hev1.4.16.L60.BC.28"_s), "main-444-10-intra");
+    ASSERT_STREQ(parseHEVCProfile("hev1.4.16.L30.B0.20"_s), "main-444-16-intra");
+
+    ASSERT_STREQ(parseHEVCProfile("hev1.5.32.L30.BE.0C"_s), "high-throughput-444");
```
Should be `hev1.5.20.L30.BE.0C`

StaZhu commented on Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
```diff
@@ -66,6 +66,43 @@ TEST_F(GStreamerTest, gstStructureJSONSerializing)
+    ASSERT_STREQ(parseHEVCProfile("hev1.5.32.L30.B0.0C"_s), "high-throughput-444-14");
+    ASSERT_STREQ(parseHEVCProfile("hev1.5.32.L30.B0.24"_s), "high-throughput-444-16-intra");
+
+    ASSERT_STREQ(parseHEVCProfile("hev1.9.512.L30.BF.8C"_s), "screen-extended-main");
```
should be `hev1.9.200.L30.BF.8C`

philn committed:
[GStreamer][WebCodecs] HEVC encoding and decoding support
https://bugs.webkit.org/show_bug.cgi?id=260124

Reviewed by Xabier Rodriguez-Calvar.

The corresponding layout tests are now passing, excepted the h265_annexb variants, that would
require further investigation. Support for H265 codec string to GStreamer H265 profile was also
added, with unit-tests.

* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/GStreamer.cmake:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::configureVideoDecoderForHarnessing):
(WebCore::wrapSpanData):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::initializeEncoders):
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::initialize):
* Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.cpp:
(WebCore::GStreamerCodecUtilities::parseHEVCProfile):
* Source/WebCore/platform/gstreamer/GStreamerCodecUtilities.h:
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
(webkit_video_encoder_class_init):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/266904@main
